[@property] Handle dynamic updates to viewport units

heycam opened the pull request:
#### bb27e4ae20f7f78a43ef1347c25be6eead488f29
<pre>
[@property] Handle dynamic updates to viewport units
<a href="https://bugs.webkit.org/show_bug.cgi?id=255689">https://bugs.webkit.org/show_bug.cgi?id=255689</a>
rdar://108287215

Reviewed by Antti Koivisto.

Viewport units are &quot;computationally independent&quot;, and so are allowed in
registered custom property initial values. But we don&apos;t have any
mechanism to invalidate them when the viewport size changes.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):

Record whether the value had any viewport units.

* Source/WebCore/css/CSSRegisteredCustomProperty.h:

Store the CSS parser tokens an initial value was created from if it
it contains viewport units, so that we can re-parse it when the viewport
is resized.

* Source/WebCore/css/CSSValue.h:
(WebCore::ComputedStyleDependencies::isComputationallyIndependent const):
(WebCore::ComputedStyleDependencies::isEmpty const): Deleted.

Rename isEmpty, since we don&apos;t want it to look at the new
viewportDimensions member.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::customPropertyValue const):
(WebCore::ComputedStyleExtractor::propertyValue const):

Flush layout in the parent document if there are viewport units used
anywhere in the current document.

* Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
(WebCore::DOMCSSRegisterCustomProperty::registerProperty):
* Source/WebCore/style/CustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::registerFromAPI):
(WebCore::Style::CustomPropertyRegistry::registerFromStylesheet):
(WebCore::Style::CustomPropertyRegistry::parseInitialValue):

Factor out initial value parsing from
DOMCSSRegisterCustomProperty::registerProperty and
CustomPropertyRegistry::registerFromStylesheet into a new function
parseInitialValue, which returns both the newly parsed
CSSCustomPropertyValue as well as whether it contained viewport unit
values.

* Source/WebCore/style/CustomPropertyRegistry.h:
(WebCore::Style::CustomPropertyRegistry::invalidatePropertiesWithViewportUnits):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::didChangeViewportSize):

Respond to viewport size changes by invalidating all registered custom
properties whose initial value contains viewport units.

Canonical link: <a href="https://commits.webkit.org/267590@main">https://commits.webkit.org/267590@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fd8ad18d12c629a6ba52528aaaa5fa99c9319e6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17081 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17406 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17889 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18867 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15981 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17273 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20680 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17548 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18198 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17284 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17644 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14822 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19683 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14884 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15513 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22215 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15883 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15680 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20032 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16267 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13799 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15424 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4079 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19789 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16100 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/style/CustomPropertyRegistry.h:
```diff
@@ -50,14 +50,20 @@ class CustomPropertyRegistry {
+    void invalidatePropertiesWithViewportUnits(Document&);
+
+    enum class ViewportUnitDependency : bool { No, Yes };
+    enum class ParseInitialValueError : uint8_t { NotComputationallyIndependent, DidNotParse };
```
It's short enough to be a `bool`.

heycam replied:
My brain thought `bool` but my fingers typed `uint8_t`. :-)

heycam replied:
Actually, I think it feels very slightly odd to be using bool for enum classes that don't represent a yes/no true/false kind of value.

mdubet replied:
We have some in the codebase, but indeed we are not consistent about 2 values "non boolean" enum, neither the coding style. I don't know what we should do about it 😅

mdubet commented on Source/WebCore/style/CustomPropertyRegistry.cpp:
```diff
@@ -194,5 +188,50 @@ void CustomPropertyRegistry::notifyAnimationsOfCustomPropertyRegistration(const
+    bool invalidatedAny = false;
+
+    auto invalidatePropertiesWithViewportUnits = [&](auto& map) {
+        for (auto& property : map.values()) {
```
We could check (or assert) for `property` before using it.

heycam replied:
We never add null to the maps. That calls for making the type `UniqueRef` instead of `unique_ptr`.

mdubet commented on Source/WebCore/style/CustomPropertyRegistry.cpp:
```diff
@@ -194,5 +188,50 @@ void CustomPropertyRegistry::notifyAnimationsOfCustomPropertyRegistration(const
+
+    // We don't need to provide a real context style since only computationally independent values are allowed (no 'em' etc).
+    auto placeholderStyle = RenderStyle::create();
+    Style::Builder builder { placeholderStyle, { document, RenderStyle::defaultStyle() }, { }, { } };
```
It seems we need a new static constructor `Style::Builder::makeWithoutStyleContext()` ?

heycam replied:
The constructor takes a reference to a `RenderStyle`, which is not a refcounted object. So we'd have to have a `RenderStyle` embedded in / owned by the `Builder` itself for cases like this. I'm going to leave this as is.

anttijk commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1743,6 +1725,26 @@ void CSSPrimitiveValue::collectComputedStyleDependencies(ComputedStyleDependenci
     case CSSUnitType::CSS_DVMAX:
     case CSSUnitType::CSS_DVB:
     case CSSUnitType::CSS_DVI:
+        dependencies.viewportDimensions = true;
```
nice!

anttijk commented on Source/WebCore/style/StyleScope.cpp:
```diff
@@ -845,6 +845,15 @@ void Scope::didChangeViewportSize()
+            if (auto element = downcast<Document>(rootNode).documentElement())
+                element->invalidateStyleForSubtree();
+        } else if (is<Element>(rootNode))
+            downcast<Element>(rootNode).invalidateStyleForSubtree();
```
`rootNode` here is always either a `ShadowRoot` or `Document`, neither of which is an `Element`. 

I don't think you need to do anything here in the shadow root case, invalidation the document by its scope is sufficient.

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1743,6 +1725,26 @@ void CSSPrimitiveValue::collectComputedStyleDependencies(ComputedStyleDependenci
     case CSSUnitType::CSS_DVMAX:
     case CSSUnitType::CSS_DVB:
     case CSSUnitType::CSS_DVI:
+        dependencies.viewportDimensions = true;
```
You can start using it at https://searchfox.org/wubkat/rev/f61c5111898cf9fd9ee106240f085c69776dc8a8/Source/WebCore/css/CSSPrimitiveValueMappings.h#1973-1974 and remove the calc() case (collectComputedStyleDependencies already traverses recursively)

heycam replied:
Could you file the bug (or write the patch)?

heycam committed:
[@property] Handle dynamic updates to viewport units
https://bugs.webkit.org/show_bug.cgi?id=255689
rdar://108287215

Reviewed by Antti Koivisto.

Viewport units are "computationally independent", and so are allowed in
registered custom property initial values. But we don't have any
mechanism to invalidate them when the viewport size changes.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):

Record whether the value had any viewport units.

* Source/WebCore/css/CSSRegisteredCustomProperty.h:

Store the CSS parser tokens an initial value was created from if it
it contains viewport units, so that we can re-parse it when the viewport
is resized.

* Source/WebCore/css/CSSValue.h:
(WebCore::ComputedStyleDependencies::isComputationallyIndependent const):
(WebCore::ComputedStyleDependencies::isEmpty const): Deleted.

Rename isEmpty, since we don't want it to look at the new
viewportDimensions member.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::customPropertyValue const):
(WebCore::ComputedStyleExtractor::propertyValue const):

Flush layout in the parent document if there are viewport units used
anywhere in the current document.

* Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
(WebCore::DOMCSSRegisterCustomProperty::registerProperty):
* Source/WebCore/style/CustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::registerFromAPI):
(WebCore::Style::CustomPropertyRegistry::registerFromStylesheet):
(WebCore::Style::CustomPropertyRegistry::parseInitialValue):

Factor out initial value parsing from
DOMCSSRegisterCustomProperty::registerProperty and
CustomPropertyRegistry::registerFromStylesheet into a new function
parseInitialValue, which returns both the newly parsed
CSSCustomPropertyValue as well as whether it contained viewport unit
values.

* Source/WebCore/style/CustomPropertyRegistry.h:
(WebCore::Style::CustomPropertyRegistry::invalidatePropertiesWithViewportUnits):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::didChangeViewportSize):

Respond to viewport size changes by invalidating all registered custom
properties whose initial value contains viewport units.

Canonical link: https://commits.webkit.org/267590@main
